File: KalmanAxioms.ml
#use "topfind";;
Topfind.don't_load ["compiler-libs.toplevel"];; 
#require "owl";;
open Owl;;
#require "owl_top";;

let x = Mat.eye 5;;

let cholesky_DC _matrixElem n mat = mat

let dense_fill _matrixElem (nc: int) (nl: int) (fn: int -> int -> float) = Mat.init_2d nc nl fn;; (*  *)

let dense_get _matrixElem (_nc: int) (_nl: int) mat l c = Mat.get mat l c;; (*  *)

let dense_elementwise_op _matrixElem (_nc: int) (_nl: int) (op: float -> float -> float) mat1 mat2 =
  dense_fill _matrixElem _nc _nl (fun x y -> op (dense_get _matrixElem _nc _nl mat1 x y) (dense_get _matrixElem _nc _nl mat2 x y));;

let dense_mul _matrixElem (_nc: int) (_nc': int) (_nl: int) mat1 mat2  = Mat.(mat1 * mat2);; (*  *)

let dense_sparse_mul _matrixElem n mat1 mat2 = Mat.(mat1 * (Owl_sparse_matrix.D.to_dense mat2));; (*  *)

let dense_sparse_mul_to_sparse _matrixElem n mat1 mat2 = Owl_sparse_matrix.D.of_dense (dense_sparse_mul _matrixElem n mat1 mat2);; (*  *)

let float_div x y = x /. y

let float_eq_dec x y = x = y

let float_inv x = 1. /. x

let float_minus x y = x -. y

let float_one = 1.0

let float_opp x = -. x

let float_plus x y = x +. y

let float_times x y = x *. y

let float_zero = 0.

let solveR_lower _matrixElem n mat1 mat2 = Mat.(Mat.inv mat1 * mat2);; (*  *)

let solveR_upper _matrixElem n mat1 mat2 = Mat.(Mat.inv mat1 * mat2);; (*  *)

let sparse_dense_mul _matrixElem n mat1 mat2 = Mat.((Owl_sparse_matrix.D.to_dense mat2) * mat1);; (*  *)

let sparse_fill _matrixElem (nc: int) (nl: int) (fn: int -> int -> float) = Owl_sparse_matrix.D.of_dense (dense_fill _matrixElem nc nl fn);; (*  *)

File: haskell/KalmanAxioms.hs
module KalmanAxioms
where

import Numeric.LinearAlgebra
import Unsafe.Coerce

type Float_t = Double
type SDM = Matrix Float_t
type SSM = Matrix Float_t

float_zero :: Float_t
float_zero = 0

float_one :: Float_t
float_one = 1

float_opp :: Float_t -> Float_t
float_opp = negate

float_plus :: Float_t -> Float_t -> Float_t
float_plus = (+)

float_minus :: Float_t -> Float_t -> Float_t
float_minus = (-)

float_times :: Float_t -> Float_t -> Float_t
float_times = (*)

float_div :: Float_t -> Float_t -> Float_t
float_div = (/)

float_inv :: Float_t -> Float_t
float_inv = (1 /)

float_eq_dec :: Float_t -> Float_t -> Bool
float_eq_dec = (==)

float_lt :: a -> a' -> Bool
float_lt x y = (unsafeCoerce x :: Float_t) < (unsafeCoerce y :: Float_t)

cholesky_DC :: a -> Int -> sdm -> sdm'
cholesky_DC _ _ a = unsafeCoerce (chol (trustSym (unsafeCoerce a :: SDM)))

dense_sparse_mul_to_sparse :: a -> Int -> sdm -> ssm -> ssm'
dense_sparse_mul_to_sparse _ _ a b = unsafeCoerce ((unsafeCoerce a :: SDM) <> (unsafeCoerce b :: SSM))

sparse_dense_mul :: a -> Int -> ssm -> sdm -> sdm'
sparse_dense_mul _ _ a b = unsafeCoerce ((unsafeCoerce a :: SSM) <> (unsafeCoerce b :: SDM))

dense_sparse_mul :: a -> Int -> sdm -> ssm -> sdm'
dense_sparse_mul _ _ a b = unsafeCoerce ((unsafeCoerce a :: SDM) <> (unsafeCoerce b :: SSM))

solveR_upper :: a -> Int -> sdm -> sdm -> sdm'
solveR_upper _ _ a b = unsafeCoerce ((unsafeCoerce a :: SDM) <\> (unsafeCoerce b :: SDM))

solveR_lower :: a -> Int -> sdm -> sdm -> sdm'
solveR_lower _ _ a b = unsafeCoerce ((unsafeCoerce a :: SDM) <\> (unsafeCoerce b :: SDM))

dense_get :: a -> Int -> Int -> sdm -> Int -> Int -> Float_t
dense_get _ _ _ m i j = (unsafeCoerce m :: SDM) `atIndex` (i, j)

dense_mul :: a -> Int -> Int -> Int -> sdm -> sdm' -> sdm''
dense_mul _ _ _ _ a b = unsafeCoerce ((unsafeCoerce a :: SDM) <> (unsafeCoerce b :: SDM))

dense_fill :: a -> Int -> Int -> (Int -> Int -> Float_t) -> sdm
dense_fill _ m n f = unsafeCoerce (fromLists [ [f i j | j <- [0..n - 1]] | i <- [0..m-1]] :: SDM)

dense_elementwise_op :: a -> Int -> Int -> (Float_t -> Float_t -> Float_t) -> SDM -> SDM -> SDM
dense_elementwise_op me m n f a b = dense_fill me m n (\i j -> f (a `atIndex` (i, j)) (b `atIndex` (i, j)))

sparse_get :: a -> Int -> Int -> SSM -> Int -> Int -> Float_t
sparse_get _ _ _ m i j = m `atIndex` (i, j)

sparse_mul :: a -> Int -> Int -> Int -> SSM -> SSM -> SSM
sparse_mul _ _ _ _ a b = a <> b

sparse_fill :: a ->  Int -> Int -> (Int -> Int -> Float_t) -> SSM
sparse_fill _ m n f = fromLists [ [f i j | j <- [0..n - 1]] | i <- [0..m-1]]

sparse_elementwise_op :: a -> Int -> Int -> (Float_t -> Float_t -> Float_t) -> SSM -> SSM -> SSM
sparse_elementwise_op me m n f a b = sparse_fill me m n (\i j -> f (a `atIndex` (i, j)) (b `atIndex` (i, j)))
